fuzz/quic-client.c

```diff
@@ -0,0 +1,121 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * https://www.openssl.org/source/license.html
+ * or in the file LICENSE in the source distribution.
+ */
+
+#include <time.h>
+#include <openssl/rand.h>
+#include <openssl/ssl.h>
+#include <openssl/rsa.h>
+#include <openssl/dsa.h>
+#include <openssl/ec.h>
+#include <openssl/dh.h>
+#include <openssl/err.h>
+#include "fuzzer.h"
+
+/* unused, to avoid warning. */
+static int idx;
+
+#define FUZZTIME 1485898104
+
+#define TIME_IMPL(t) { if (t != NULL) *t = FUZZTIME; return FUZZTIME; }
+
+/*
+ * This might not work in all cases (and definitely not on Windows
+ * because of the way linkers are) and callees can still get the
+ * current time instead of the fixed time. This will just result
+ * in things not being fully reproducible and have a slightly
+ * different coverage.
+ */
+#if !defined(_WIN32)
+time_t time(time_t *t) TIME_IMPL(t)
+#endif
+
+int FuzzerInitialize(int *argc, char ***argv)
+{
+    STACK_OF(SSL_COMP) *comp_methods;
+
+    FuzzerSetRand();
+    OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS | OPENSSL_INIT_ASYNC, NULL);
+    OPENSSL_init_ssl(OPENSSL_INIT_LOAD_SSL_STRINGS, NULL);
+    ERR_clear_error();
+    CRYPTO_free_ex_index(0, -1);
+    idx = SSL_get_ex_data_X509_STORE_CTX_idx();
+    comp_methods = SSL_COMP_get_compression_methods();
+    if (comp_methods != NULL)
+        sk_SSL_COMP_sort(comp_methods);
+
+    return 1;
+}
+
+int FuzzerTestOneInput(const uint8_t *buf, size_t len)
+{
+    SSL *client = NULL;
+    BIO *in;
+    BIO *out;
+    SSL_CTX *ctx;
+
+    if (len == 0)
+        return 0;
+
+    /* This only fuzzes the initial flow from the client so far. */
+    ctx = SSL_CTX_new(OSSL_QUIC_client_method());
+    if (ctx == NULL)
+        goto end;
+
+    client = SSL_new(ctx);
+    if (client == NULL)
+        goto end;
+/*
+    OPENSSL_assert(SSL_set_min_proto_version(client, 0) == 1);
+    OPENSSL_assert(SSL_set_cipher_list(client, "ALL:eNULL:@SECLEVEL=0") == 1);
+*/
+    SSL_set_tlsext_host_name(client, "localhost");
+    in = BIO_new(BIO_s_dgram_mem());
+    if (in == NULL)
+        goto end;
+    out = BIO_new(BIO_s_dgram_mem());
+    if (out == NULL) {
+        BIO_free(in);
+        goto end;
+    }
+    if (SSL_set_alpn_protos(client, (const unsigned char *)"\x08quicfuzz", 9) != 0)
+        goto end;
+    SSL_set_bio(client, in, out);
+    SSL_set_connect_state(client);
+    while (len > 3)
+    {
+        size_t size = buf[0] + (buf[1] << 8);
+
+        if (size > len - 2)
+            break;
+
+        if (size > 0)
+            OPENSSL_assert((size_t)BIO_write(in, buf+2, size) == size);
+        len -= size + 2;
+        buf += size + 2;
+
+        if (SSL_do_handshake(client) == 1) {
+            /* Keep reading application data until error or EOF. */
+            uint8_t tmp[1024];
+            if (SSL_read(client, tmp, sizeof(tmp)) <= 0)
+                break;
+        }
```

Comment:
You should probably call `SSL_tick()` somewhere here.

From the documentation:

> By default, blocking is enabled, unless the SSL object is
configured to use an underlying read or write BIO which cannot provide a poll
descriptor (see L<BIO_get_rpoll_descriptor(3)>), as blocking mode cannot be
supported in this case.

`BIO_s_dgram_mem()` does not support `BIO_get_rpoll_descriptor()` so we will default to non-blocking mode. Then:

> Where a QUIC connection SSL object is used in nonblocking mode, an application
is responsible for ensuring that the SSL object is ticked regularly; see
L<SSL_tick(3)>.